include/Beman/Optional26/optional.hpp

```diff
@@ -950,209 +1106,466 @@ auto optional_map_impl(Opt&& opt, F&& f)
+    constexpr optional& operator=(optional&& rhs) noexcept      = default;
 
     template <class U = T>
-        requires(!detail::is_optional<std::decay_t<U>>)
-    constexpr explicit(!std::is_convertible_v<U, T>) optional(U&& u) noexcept : value_(std::addressof(u)) {
-        static_assert(std::is_constructible_v<std::add_lvalue_reference_t<T>, U>, "Must be able to bind U to T&");
-        static_assert(std::is_lvalue_reference<U>::value, "U must be an lvalue");
-    }
+        requires(!detail::is_optional<decay_t<U>>)
+    constexpr optional& operator=(U&& u);
 
     template <class U>
-    constexpr explicit(!std::is_convertible_v<U, T>) optional(const optional<U>& rhs) noexcept {
-        static_assert(std::is_constructible_v<std::add_lvalue_reference_t<T>, U>, "Must be able to bind U to T&");
-        if (rhs.has_value())
-            value_ = std::to_address(rhs);
-        else
-            value_ = nullptr;
-    }
+    constexpr optional& operator=(const optional<U>& rhs) noexcept;
 
-    //  \rSec3[optional.dtor]{Destructor}
+    template <class U>
+    constexpr optional& operator=(optional<U>&& rhs) = delete;
 
-    constexpr ~optional() = default;
+    template <class U>
+        requires(!detail::is_optional<decay_t<U>>)
+    constexpr optional& emplace(U&& u) noexcept;
+
+    // \ref{optionalref.swap}, swap
+    constexpr void swap(optional& rhs) noexcept;
+
+    // \ref{optional.iterators}, iterator support
+    constexpr iterator       begin() noexcept;
+    constexpr const_iterator begin() const noexcept;
+    constexpr iterator       end() noexcept;
+    constexpr const_iterator end() const noexcept;
+
+    // \ref{optionalref.observe}, observers
+    constexpr T*       operator->() const noexcept;
+    constexpr T&       operator*() const noexcept;
+    constexpr explicit operator bool() const noexcept;
+    constexpr bool     has_value() const noexcept;
+    constexpr T&       value() const;
+    template <class U>
+    constexpr T value_or(U&& u) const;
+
+    // \ref{optionalref.monadic}, monadic operations
+    template <class F>
+    constexpr auto and_then(F&& f) const;
+    template <class F>
+    constexpr auto transform(F&& f) const -> optional<invoke_result_t<F, T&>>;
+    template <class F>
+    constexpr optional or_else(F&& f) const;
+
+    // \ref{optional.mod}, modifiers
+    constexpr void reset() noexcept;
+
+  private:
+    T* value_; // exposition only
+};
 
-    // \rSec3[optional.assign]{Assignment}
+//  \rSec3[optionalref.ctor]{Constructors}
+template <class T>
+constexpr optional<T&>::optional() noexcept : value_(nullptr) {}
+
+template <class T>
+constexpr optional<T&>::optional(nullopt_t) noexcept : value_(nullptr) {}
+
+template <class T>
+template <class U>
+    requires(!detail::is_optional<decay_t<U>>)
+constexpr optional<T&>::optional(U&& u) noexcept : value_(addressof(u)) {
+    static_assert(is_constructible_v<add_lvalue_reference_t<T>, U>, "Must be able to bind U to T&");
```

Comment:
Why are `add_lvalue_reference_t` and `add_rvalue_reference_t` used given `T&` must be well-formed in these partial specializations?